lib/mongo/server/connection_pool/queue.rb

```diff
@@ -79,6 +83,10 @@ def initialize(options = {}, &block)
           @queue = Array.new(min_size) { create_connection }
           @mutex = Mutex.new
           @resource = ConditionVariable.new
+
+          @wait_queue = []
+          @wait_queue_mutex = Mutex.new
```

Comment:
Why are there now two parallel lock implementations?

Reply:
One is for properly handling access to the list of available connections (i.e. `@queue`) and one is for properly handling access to the list of waiting threads (i.e. `@wait_queue)

Reply:
I believe this has been resolved by our offline discussions and the recent updates